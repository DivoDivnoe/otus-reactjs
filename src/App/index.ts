export { App, AppRoutes } from './App';
