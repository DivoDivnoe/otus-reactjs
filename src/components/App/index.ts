export { App, AppProps, AppRoutes } from './App';
