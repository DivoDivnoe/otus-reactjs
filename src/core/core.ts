import { getMatrixItemNeighbours } from '@/utils';
import { getRandomValuesArr, getZeroMatrix } from '@/utils';
import { BoardSize, FillType, CellState } from '@/constants';
import { BoardSizeValue, BoardFillPercentage } from '@/configs';

export type Binary = 0 | 1;
export type Model = Binary[][];

export interface SizeProps {
  width: number;
  height: number;
}

export const GameCoreLogicConfig = {
  NEIGHBOURS_AMOUNT_TO_BORN_NEW_ITEM: 3,
  MIN_NEIGHBOURS_AMOUNT_TO_STAY_ALIVE: 2,
  MAX_NEIGHBOURS_AMOUNT_TO_STAY_ALIVE: 3,
};

export const getRandomMatrix = (size: SizeProps, fill: number): Model => {
  const { width, height } = size;
  const maxValue = width * height;
  const amount = Math.round(fill * maxValue);
  const randomIndexes = getRandomValuesArr(maxValue, amount);

  return Array.from({ length: height }, (_, rowIndex) => {
    return Array.from({ length: width }, (_, columnIndex) => {
      return randomIndexes.includes(rowIndex * width + columnIndex)
        ? CellState.ALIVE
        : CellState.DEAD;
    });
  });
};

export const createRandomMatrix = (
  sizeType: BoardSize,
  fillType: FillType
): Model => {
  const size = BoardSizeValue[sizeType];
  const fill = BoardFillPercentage[fillType];

  return getRandomMatrix(size, fill);
};

export const getNewSizeMatrix = (size: SizeProps, prevMatrix: Model): Model => {
  const prevMatrixHeight = prevMatrix.length;
  const prevMatrixWidth = prevMatrix[0].length;
  const { width, height } = size;

  if (width === prevMatrixWidth && height === prevMatrixHeight)
    return prevMatrix;

  return Array.from({ length: height }, (_, rowIndex) => {
    return Array.from({ length: width }, (_, columnIndex) => {
      const prevItem = prevMatrix[rowIndex]?.[columnIndex];

      return prevItem === undefined ? CellState.DEAD : prevItem;
    });
  });
};

export const createNewSizeMatrix = (
  sizeType: BoardSize,
  prevMatrix: Model
): Model => {
  const size = BoardSizeValue[sizeType];

  return getNewSizeMatrix(size, prevMatrix);
};

export const createZeroMatrix = (sizeType: BoardSize): Model => {
  const size = BoardSizeValue[sizeType];

  return getZeroMatrix(size) as Model;
};

export const getNextGenMatrix = (prevGenMatrix: Model): Model => {
  const {
    MIN_NEIGHBOURS_AMOUNT_TO_STAY_ALIVE,
    MAX_NEIGHBOURS_AMOUNT_TO_STAY_ALIVE,
    NEIGHBOURS_AMOUNT_TO_BORN_NEW_ITEM,
  } = GameCoreLogicConfig;

  return Array.from({ length: prevGenMatrix.length }, (_, rowIndex) => {
    return Array.from(
      { length: prevGenMatrix[rowIndex].length },
      (_, columnIndex) => {
        const currentItem = prevGenMatrix[rowIndex][columnIndex];
        const aliveNeighboursLength = getMatrixItemNeighbours(
          prevGenMatrix,
          rowIndex,
          columnIndex
        ).filter((item) => item === CellState.ALIVE).length;

        if (
          currentItem === CellState.DEAD &&
          aliveNeighboursLength === NEIGHBOURS_AMOUNT_TO_BORN_NEW_ITEM
        ) {
          return CellState.ALIVE;
        }

        if (
          currentItem === CellState.ALIVE &&
          (aliveNeighboursLength < MIN_NEIGHBOURS_AMOUNT_TO_STAY_ALIVE ||
            aliveNeighboursLength > MAX_NEIGHBOURS_AMOUNT_TO_STAY_ALIVE)
        ) {
          return CellState.DEAD;
        }

        return currentItem;
      }
    );
  });
};
