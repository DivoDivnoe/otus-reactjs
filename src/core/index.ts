export { getNextGenMatrix, GameCoreLogicConfig } from './core';
