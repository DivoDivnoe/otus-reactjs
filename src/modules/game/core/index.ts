export {
  getNextGenMatrix,
  GameCoreLogicConfig,
  getRandomMatrix,
  createRandomMatrix,
  createZeroMatrix,
} from './core';
