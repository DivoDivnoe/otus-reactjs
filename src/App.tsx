import React, { FC } from 'react';
import { Router, Switch, Route } from 'react-router-dom';
import history from '@/history';
import { ErrorBoundary } from '@/components/ErrorBoundary';
import { PrivateRoute } from '@/modules/PrivateRoute';
import { GameScreen } from '@/screens/GameScreen';
import { SigninScreen } from '@/screens/SigninScreen';

export const App: FC = () => {
  return (
    <ErrorBoundary>
      <Router history={history}>
        <Switch>
          <Route path='/login'>
            <SigninScreen />
          </Route>
          <PrivateRoute path='/'>
            <GameScreen />
          </PrivateRoute>
        </Switch>
      </Router>
    </ErrorBoundary>
  );
};
