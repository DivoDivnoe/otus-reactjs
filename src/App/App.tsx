import React, { FC, useEffect, useCallback } from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  useHistory,
} from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { ErrorBoundary } from '@/components/ErrorBoundary';
import { StartPopup } from '@/modules/user/StartPopup';
import { PrivateRoute } from '@/components/PrivateRoute';
import { State } from '@/reducer';
import {
  getUser,
  UserType,
  ActionCreator as UserActionCreator,
} from '@/modules/user';
import { GameScreen } from '@/screens/GameScreen';
import { SigninScreen } from '@/screens/SigninScreen';

export const AppRoutes: FC = () => {
  const dispatch = useDispatch();

  const user = useSelector<State, UserType>(getUser);
  const signout = useCallback(() => dispatch(UserActionCreator.signout()), []);
  const history = useHistory();

  useEffect(() => {
    if (user) {
      history.push('/');
    } else {
      history.push('/login');
    }
  }, [user]);

  return (
    <Switch>
      <Route path='/login'>
        <SigninScreen />
      </Route>
      <PrivateRoute path='/' user='user'>
        <GameScreen user={user} signout={signout} />
      </PrivateRoute>
    </Switch>
  );
};

export const App: FC = () => {
  return (
    <ErrorBoundary>
      <Router>
        <AppRoutes />
      </Router>
    </ErrorBoundary>
  );
};
