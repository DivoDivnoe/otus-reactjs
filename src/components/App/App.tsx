import React, { FC } from 'react';
import styled from '@emotion/styled';
import { Field } from '@/components/Field';
import { ErrorBoundary } from '@/components/ErrorBoundary';
import { Bar } from '@/components/Bar/Bar';
import { StartPopup } from '@/components/StartPopup/';
import withGameLogicHOC, {
  LogicProps,
  WithGameProps,
} from '@/hocs/withGameLogicHOC';
import withUserDataHOC, { WithUserProps } from '@/hocs/withUserDataHOC';

export interface AppLogicAndUserProps extends WithUserProps, LogicProps {}
export interface AppProps extends WithGameProps, WithUserProps {}

const Title = styled.h1`
  color: #ffffff;
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const App: FC<AppProps> = (props) => {
  const {
    size,
    speed,
    fill,
    sizes,
    speedTypes,
    fillTypes,
    isPlaying,
    model,
    clickHandler,
    changeSizeHandler,
    changeSpeedHandler,
    changeFillType,
    play,
    pause,
    clear,
    user,
    setUser,
  } = props;

  return (
    <ErrorBoundary>
      <Wrapper>
        {user && <Title>Hello, {user}.</Title>}
        <Field size={size} model={model} clickHandler={clickHandler} />
        <Bar
          sizes={sizes}
          speedTypes={speedTypes}
          fillTypes={fillTypes}
          size={size}
          speed={speed}
          fill={fill}
          isPlaying={isPlaying}
          changeSizeHandler={changeSizeHandler}
          changeSpeedHandler={changeSpeedHandler}
          changeFillType={changeFillType}
          play={play}
          pause={pause}
          clear={clear}
        />
      </Wrapper>
      {!user && <StartPopup submitHandler={setUser} />}
    </ErrorBoundary>
  );
};

const AppWithGameLogic = withGameLogicHOC<AppLogicAndUserProps>(App);
const AppWithGameLogicAndUserData =
  withUserDataHOC<LogicProps>(AppWithGameLogic);
export default AppWithGameLogicAndUserData;
