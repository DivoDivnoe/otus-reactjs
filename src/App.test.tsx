import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { act } from 'react-dom/test-utils';
import { App } from './App';
import { createZeroMatrix } from '@/modules/game/core';
import { BoardSize } from './modules/game/size';
import reducer from '@/reducer';

describe('App component', () => {
  it('renders correctly not signed in', () => {
    const store = configureStore({ reducer });

    const { asFragment } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly signed in', () => {
    const store = configureStore({ reducer });
    const model = createZeroMatrix(BoardSize.MEDIUM);

    act(() => {
      store.dispatch({ type: 'model/setModel', payload: model });
      store.dispatch({ type: 'user/signin', payload: 'Andrey' });
    });

    const { asFragment } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
