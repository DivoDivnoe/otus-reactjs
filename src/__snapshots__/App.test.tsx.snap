// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly not signed in 1`] = `
<DocumentFragment>
  <div
    class="css-1i7ph3n"
  >
    <div
      class="css-1p0rybr"
      data-testid="start-popup"
    >
      <form
        class="css-qztj7d"
        data-testid="start-form"
      >
        <input
          class="css-1oh381d"
          data-testid="name-input"
          placeholder="Enter Your Name:"
          required=""
          value=""
        />
        <button
          class="css-som8ht"
        >
          Start
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App component renders correctly signed in 1`] = `
<DocumentFragment>
  <div
    class="css-1i7ph3n"
  >
    <div
      class="css-1p0rybr"
      data-testid="start-popup"
    >
      <form
        class="css-qztj7d"
        data-testid="start-form"
      >
        <input
          class="css-1oh381d"
          data-testid="name-input"
          placeholder="Enter Your Name:"
          required=""
          value=""
        />
        <button
          class="css-som8ht"
        >
          Start
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
